beacon_node/eth1/src/cache.rs

```diff
@@ -0,0 +1,182 @@
+    )
+}
+
+/// Fetches Eth1 data from the Eth1Data fetcher object.
+fn fetch_eth1_data<F: Eth1DataFetcher>(
+    distance: u64,
+    current_block_number: U256,
+    fetcher: Arc<F>,
+) -> impl Future<Item = Result<(U256, Eth1Data)>, Error = Error> + Send {
+    let block_number: U256 = current_block_number
+        .checked_sub(distance.into())
+        .unwrap_or(U256::zero());
+    let deposit_root = fetcher.get_deposit_root(Some(BlockNumber::Number(block_number.as_u64())));
+    let deposit_count = fetcher.get_deposit_count(Some(BlockNumber::Number(block_number.as_u64())));
+    let block_hash = fetcher.get_block_hash_by_height(block_number.as_u64());
+    let eth1_data_future = deposit_root.join3(deposit_count, block_hash);
+    eth1_data_future.map(move |data| {
+        let eth1_data = Eth1Data {
+            deposit_root: data.0,
+            deposit_count: data.1?,
+            block_hash: data.2.ok_or(Error::DecodingError)?,
+        };
+        Ok((block_number, eth1_data))
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::web3_fetcher::Web3DataFetcher;
+    use std::time::{Duration, Instant};
+    use tokio::timer::{Delay, Interval};
+
+    // Note: Running tests using ganache-cli instance with config
+    // from https://github.com/ChainSafe/lodestar#starting-private-eth1-chain
+
+    fn setup() -> Web3DataFetcher {
+        let deposit_contract_address = "8c594691C0E592FFA21F153a16aE41db5beFcaaa";
+        let mut abi_path = std::env::current_dir().unwrap();
+        abi_path.push("deposit_contract.json");
+        let w3 = Web3DataFetcher::new("ws://localhost:8545", deposit_contract_address, abi_path);
+        return w3.unwrap();
+    }
+
+    #[test]
+    fn test_fetch() {
+        let w3 = setup();
+        let when = Instant::now() + Duration::from_millis(5000);
+        let task1 = Delay::new(when)
+            .and_then(|_| {
+                println!("Hello world!");
+                Ok(())
+            })
+            .map_err(|e| panic!("delay errored; err={:?}", e));
+        tokio::run(task1);
+        let task2 = fetch_eth1_data(0, 10.into(), Arc::new(w3)).and_then(|data| {
+            println!("{:?}", data);
+            Ok(())
+        });
+        tokio::run(task2.map_err(|e| println!("Some error {:?}", e)));
+    }
+
+    #[test]
+    fn test_cache() {
+        let w3 = setup();
+        let interval = {
+            let update_duration = Duration::from_secs(15);
+            Interval::new(Instant::now(), update_duration).map_err(|e| println!("{:?}", e))
+        };
+
+        let cache = Eth1DataCache::new(Arc::new(w3));
+        let cache_inside = cache.cache.clone();
+        let task = interval.take(100).for_each(move |_| {
+            let _c = cache_inside.clone();
+            cache
+                .update_cache(3 + 1)
+                .and_then(move |_| Ok(()))
+                .map_err(|e| println!("Some error {:?}", e))
```

Comment:
Should this `println` be an `assert`?

If there was an error I assume we want to fail the test? I could be wrong.